File: python/zensical/compat/literate_nav.py
# Copyright (c) 2025-2026 Zensical and contributors

# SPDX-License-Identifier: MIT
# All contributions are certified under the DCO

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to
# deal in the Software without restriction, including without limitation the
# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
# sell copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.

"""Narrow Python-Markdown adapter for native literate navigation."""

from __future__ import annotations

from copy import deepcopy
from itertools import dropwhile
from typing import TYPE_CHECKING
from xml.etree import ElementTree

from markdown import Markdown
from markdown.preprocessors import Preprocessor
from markdown.treeprocessors import Treeprocessor

from zensical.config import get_config

if TYPE_CHECKING:
    from collections.abc import Callable

# ----------------------------------------------------------------------------
# Constants
# ----------------------------------------------------------------------------

# Private-use scalar framing text that must survive XML and HTML decoding.
_TEXT_ESCAPE = "\U000f0000"

# ----------------------------------------------------------------------------
# Classes
# ----------------------------------------------------------------------------


class _MarkerPreprocessor(Preprocessor):
    """Replace explicit navigation markers with a tree-visible placeholder."""

    def __init__(self, md: Markdown):
        super().__init__(md)
        self.placeholder: str | None = None

    def run(self, lines: list[str]) -> list[str]:
        for index, line in enumerate(lines):
            if line.strip() == "<!--nav-->":
                self.placeholder = self.md.htmlStash.store("")
                lines[index] = self.placeholder + "\n"
        return lines


class _CaptureTreeprocessor(Treeprocessor):
    """Capture the selected root list at literate-nav's processing phase."""

    def __init__(self, md: Markdown, marker: _MarkerPreprocessor):
        super().__init__(md)
        self.marker = marker
        self.nav: ElementTree.Element | None = None

    def run(self, root: ElementTree.Element) -> None:
        if self.marker.placeholder is None:
            candidates = reversed(root)
        else:
            candidates = dropwhile(
                lambda element: element.text != self.marker.placeholder,
                root,
            )
        for element in candidates:
            if element.tag in {"ul", "ol"}:
                self.nav = deepcopy(element)
                return


# ----------------------------------------------------------------------------
# Functions
# ----------------------------------------------------------------------------


def render(content: str) -> str:
    """Capture literate navigation with its local Markdown configuration.

    This returns only the selected list subtree. Interpretation belongs to the
    native compatibility module, keeping filesystem and navigation semantics
    out of Python.
    """
    plugin = get_config()["plugins"]["literate_nav"]["config"]
    md = Markdown(
        extensions=plugin["markdown_extensions"],
        extension_configs=plugin["mdx_configs"],
        tab_length=plugin["tab_length"],
    )

    # Parse the complete document so definitions and extension state before
    # an explicit marker remain available to the selected navigation list.
    # Keep inline HTML and entities in the captured tree instead of replacing
    # them with placeholders that only a complete Markdown render can restore.
    md.inlinePatterns.deregister("html", strict=False)
    md.inlinePatterns.deregister("entity", strict=False)
    marker = _MarkerPreprocessor(md)
    capture = _CaptureTreeprocessor(md, marker)
    md.preprocessors.register(marker, "zensical_literate_nav_marker", 25)
    md.treeprocessors.register(capture, "zensical_literate_nav_capture", 19)
    md.convert(content)
    if capture.nav is None:
        return ""
    _encode_tree(capture.nav, md.treeprocessors["unescape"].unescape)
    return ElementTree.tostring(capture.nav, encoding="unicode")


def _escape_text(value: str) -> str:
    """Encode ampersands and the escape scalar without ambiguity."""
    return value.replace(_TEXT_ESCAPE, _TEXT_ESCAPE + "S").replace(
        "&", _TEXT_ESCAPE + "A"
    )


def _encode_tree(
    root: ElementTree.Element, unescape: Callable[[str], str]
) -> None:
    """Preserve text across XML serialization and Rust's HTML tokenizer."""
    for element in root.iter():
        if element.text:
            element.text = _escape_text(unescape(element.text))
        if element.tail:
            element.tail = _escape_text(unescape(element.tail))

File: crates/zensical/src/compat/mkdocs/plugin/literate_nav.rs
// Copyright (c) 2025-2026 Zensical and contributors

// SPDX-License-Identifier: MIT
// All contributions are certified under the DCO

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to
// deal in the Software without restriction, including without limitation the
// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
// sell copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
// IN THE SOFTWARE.

// ----------------------------------------------------------------------------

//! Native compatibility pipeline for filesystem-backed literate navigation.

use anyhow::Context;
use std::collections::BTreeMap;
use std::fs;
use std::sync::Arc;

use zrx::id::Id;
use zrx::stream::function::Collection;
use zrx::stream::{Key, Signal, Stream, Value};

use crate::config::Config;
use crate::path::SourcePath;
use crate::structure::nav::{Navigation, NavigationItem, NavigationResolution};
use crate::structure::page::Page;
use crate::watcher::Source;

mod parser;
mod resolver;

// ----------------------------------------------------------------------------
// Structs
// ----------------------------------------------------------------------------

/// Literate navigation pipeline.
#[derive(Clone, Debug)]
pub struct LiterateNav {
    settings: Arc<Settings>,
}

/// Inputs required to derive revision-complete navigation.
pub struct Dependencies<'a> {
    /// Physical sources, including navigation control files.
    pub sources: &'a Stream<Id, Source>,
    /// Rendered documentation pages.
    pub pages: &'a Stream<Id, Page>,
}

/// Immutable native plugin settings.
#[derive(Clone, Debug)]
struct Settings {
    /// Whether native literate navigation participates in resolution.
    enabled: bool,
    /// Provider context containing documentation sources.
    docs: String,
    /// Folder-relative navigation document name or path.
    nav_file: String,
    /// Whether a folder index omitted from its list is inserted first.
    implicit_index: bool,
    /// Explicit MkDocs navigation used as the root fallback or seed.
    configured: Vec<NavigationItem>,
}

/// One discovered literate navigation document.
#[derive(Clone, Debug, PartialEq, Eq)]
struct Document {
    /// Canonical source-relative navigation document path.
    path: SourcePath,
    /// Complete Markdown source after optional byte-order-mark removal.
    content: String,
}

impl Value for Document {}

/// Revision-complete navigation documents.
#[derive(Clone, Debug, Default)]
struct Documents(
    /// Documents keyed by canonical source-relative path.
    Arc<BTreeMap<String, String>>,
);

impl Value for Documents {}

/// Revision-complete rendered pages.
#[derive(Clone, Debug)]
struct Pages(
    /// Rendered pages in the settled workflow revision.
    Arc<Vec<Page>>,
);

impl Value for Pages {}

// ----------------------------------------------------------------------------
// Implementations
// ----------------------------------------------------------------------------

impl LiterateNav {
    /// Resolves plugin settings for one workflow lifetime.
    pub fn new(config: &Config) -> Self {
        let plugin = &config.project.plugins.literate_nav.config;
        Self {
            settings: Arc::new(Settings {
                enabled: plugin.enabled,
                docs: config.project.docs_dir.clone(),
                nav_file: plugin.nav_file.clone(),
                implicit_index: plugin.implicit_index,
                configured: config.project.nav.clone(),
            }),
        }
    }

    /// Installs navigation discovery, settlement, and compilation.
    pub fn setup(
        &self, dependencies: Dependencies<'_>,
    ) -> Signal<Id, NavigationResolution> {
        let settings = self.settings.clone();
        let documents = dependencies.sources.filter_map({
            let settings = settings.clone();
            move |id: &Id, source: &Source| {
                if !settings.enabled || id.context() != settings.docs {
                    return Ok(None);
                }
                let path = id.location().parse::<SourcePath>()?;
                if !is_navigation_file(&path, &settings.nav_file) {
                    return Ok(None);
                }
                let content =
                    fs::read_to_string(&**source).with_context(|| {
                        format!(
                            "failed to read literate navigation file {path}"
                        )
                    })?;
                Ok::<_, anyhow::Error>(Some(Document {
                    path,
                    content: content.trim_start_matches('\u{feff}').into(),
                }))
            }
        });
        let documents = documents.reduce(
            |documents: &dyn Collection<Key<Id>, Document>| {
                Some(Documents(Arc::new(
                    documents
                        .values()
                        .map(|document| {
                            (
                                document.path.to_string(),
                                document.content.clone(),
                            )
                        })
                        .collect(),
                )))
            },
        );
        let pages = dependencies.pages.reduce(
            |pages: &dyn Collection<Key<Id>, Page>| {
                Some(Pages(Arc::new(pages.values().cloned().collect())))
            },
        );

        let navigation = pages.product(&documents).map(
            move |pages: &Pages, docs: &Documents| {
                if settings.enabled {
                    resolver::resolve(&settings, &docs.0, pages.0.as_ref())
                } else {
                    Ok(Navigation::resolve(
                        settings.configured.clone(),
                        pages.0.as_ref().clone(),
                    ))
                }
            },
        );
        navigation.reduce(
            |navigation: &dyn Collection<Key<Id>, NavigationResolution>| {
                navigation.values().next().cloned()
            },
        )
    }
}

/// Returns whether a source can serve as a folder navigation document.
fn is_navigation_file(path: &SourcePath, nav_file: &str) -> bool {
    path.as_str() == nav_file
        || path
            .as_str()
            .strip_suffix(nav_file)
            .is_some_and(|prefix| prefix.ends_with('/'))
}

// ----------------------------------------------------------------------------
// Tests
// ----------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::is_navigation_file;

    #[test]
    fn discovers_simple_and_nested_navigation_names() {
        assert!(is_navigation_file(
            &"guide/SUMMARY.md".parse().unwrap(),
            "SUMMARY.md"
        ));
        assert!(is_navigation_file(
            &"guide/nav/SUMMARY.md".parse().unwrap(),
            "nav/SUMMARY.md"
        ));
        assert!(!is_navigation_file(
            &"guide/NOT-SUMMARY.md".parse().unwrap(),
            "SUMMARY.md"
        ));
    }
}
